### ISSUE TITLE
Handling of ties during evaluation.

### ISSUE THREAD

**Issue by sfschouten** [uma-pi1/kge#221]
TLDR; I have been observing very high performance for a very simple model, I think a bug in the evaluation may be the cause.

I have been experimenting with learning embeddings for entities' types.
The simplest application of this is to model an entity's embedding as the mean of its types' embeddings. 
For the WN18RR dataset, this means that each entity's embedding is always that entity's type's embedding (of which there is only one for that dataset), this results in the same score being assigned to many triples.
This setting resulted in MRR scores of up to 0.723, which seems suspiciously high. And this is using the rounded_mean_rank tie handling strategy.

I have been logging a lot of values and I believe I have found the cause. In the following code snippet, floating-point issues are supposedly dealt with. 
https://github.com/uma-pi1/kge/blob/2ecac7fa8fd743f231519ad059665488440bc616/kge/job/eval_entity_ranking.py#L214-L221

However, it seems that the code is performing the opposite assignment that it should be doing. Even the comment seems to suggest the order should have been the opposite.

The way it is right now, it seems like the number of ties is pretty much always 1. This is because the value in true_scores will always match its own value, but small floating-point differences are causing it to never match the values in the chunk-based tensor of scores.
This means that ties are effectively decided 'randomly'. This does leave the question of how this 'random' tie-breaking then results in high MRR (even on the test set), but one step at a time (perhaps this tie-breaking is being 'tuned' during the hyperparameter search).

I think the code should be changed to:
```Python
                o_true_scores[o_in_chunk_mask] = scores_sp[o_in_chunk_mask, o_in_chunk]
                s_true_scores[s_in_chunk_mask] = scores_po[s_in_chunk_mask, s_in_chunk]
```
This results in the number of ties I expect for my model.

Let me know if this makes sense, more than happy to open a PR if it does.

**Comment by rgemulla** [comment]
Thank you for reporting this issue. 

The main reason for the floating point issues is that (i) the true scores are computed via SPO scoring and (ii) the negative scores are computed via SP or PO scoring. This former is necessary so that the true scores are known upfront to perform evaluation in chunks, the latter is necessary to achieve reasonable efficiency.

That being said, the problem that you mention can indeed occur and it would be good to fix it. The suggested fix is problematic, however, since it produces inconsistent results when the entity ranking is performed in chunks (as is necessary for large datasets / low-memory GPUs). This is because the "true scores" of an entity are used right away (starting from the first chunk), whereas the code that you give may updates them later and thus may invalidate the ranks computed so far.

One workable solution may be to create a "special chunk" upfront which consist only of the positives, compute all SP/PO scores and use the result to initialize s_true_scores and o_true_scores. This will slow down evaluation, but give more consistent results.

@sfschouten @AdrianKs Any thoughts on this?

**Comment by sfschouten** [comment]
Right! I see now why my solution is problematic. I think the idea of doing a 'special' chunk first makes a lot of sense. And I think the slow down will probably be very manageable.

If not, another solution might be to require models to have an option that switches on a 'consistent' mode, where the same function is used for SPO scoring as for SP or PO scoring. This consistent mode could then be used specifically during evaluation.
This would leave performance the same for models that are 'consistent' already.

**Comment by AdrianKs** [comment]
I think we could indeed just use SP and PO functions instead of SPO to calculate the positives. For this we would need to get unique Os to then do SP_uniqueOs and the other way around. Afterwards, we filter for the scores of the triples we are interested in.

**Comment by rufex2001** [comment]
@AdrianKs makes sense! As done in NegSamp! Might even be faster sometimes

**Comment by sfschouten** [comment]
I have discovered one more way this can go wrong.
The `torch.cdist` function has a `compute_mode` which by default is set to `'use_mm_for_euclid_dist_if_necessary'`. Apparently, this means the function uses the size of the tensors passed to it to determine if it will use matrix multiplication to approximate the Euclidean distance. I have now noticed cases where the calls for the true_scores are using this matrix multiplication based approximation and the call for the other scores are not, leading to the same bug.

Perhaps the score_emb of each method should just be changed to be completely deterministic during evaluation, if model.training() returns false then no input-dependent behaviour at all?

**Comment by rgemulla** [comment]
I had thought of such issues as well; e.g., operations such as matrix multiply may use different implementations depending on inputs. Perhaps a better solution is to use `torch.isclose` to check for ties.

@sfschouten: would this work in your setting?

**Comment by sfschouten** [comment]
Yes, I am starting to think this may be the best option.
Of course, this does make it possible for scores that were very close but not actually equal to be treated as though they were. 
But it's probably better to detect ties where there are none than to miss them when they do exist. Perhaps the tolerances should be configurable?

**Comment by AdrianKs** [comment]
`torch.isclose()` allows to configure absolute or relative tolerance. So this is definitely possible to make it configurable.
